### ISSUE TITLE
Generated X.509 certs should be RFC 5280 compliant by default

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#13603]
Meanwhile we have stronger checks for X.509 certificates to comply to RFC 5280, 
at least when strict checking is enabled (e.g., using `-x509_strict`).

Yet unfortunately the OpenSSL apps by default tend to generate certs that are not compliant.
In particular, X.509v3 certs (i.e., those having any X.509 extensions) MUST include 
*  an authority key identifier (AKID) with the exception of self-signed certs - see https://tools.ietf.org/html/rfc5280#section-4.2.1.1
* a subject key identifier (SKID) with the exception of non-CA certs - see https://tools.ietf.org/html/rfc5280#section-4.2.1.2

**Comment by t8m** [comment]
I suppose this is something in between bug and feature so I do not expect we will fix this in 1.1.1.

**Comment by t8m** [comment]
This is something that would be nice to get fixed for 3.0 however it is IMO not a blocker for 3.0 either.

**Comment by ronf** [comment]
Regarding non-CA certs, the subject key identifier (SKID) is still listed as a SHOULD in RFC 5280:

>    For end entity certificates, the subject key identifier extension
>    provides a means for identifying certificates containing the
>    particular public key used in an application.  Where an end entity
>    has obtained multiple certificates, especially from multiple CAs, the
>    subject key identifier provides a means to quickly identify the set
>    of certificates containing a particular public key.  To assist
>    applications in identifying the appropriate end entity certificate,
>    this extension SHOULD be included in all end entity certificates.

So, while not mandatory, we might want OpenSSL apps generating end-entity certs to include a SKID by default (along with an AKID if they are not self-signed).

**Comment by richsalz** [comment]
The strict cert checking should be fixed to warn about SKID, as in the quote above from RFC 5280.  Want me to open a separate issue?

**Comment by DDvO** [comment]
> The strict cert checking should be fixed to warn about SKID, as in the quote above from RFC 5280. Want me to open a separate issue?

I support this, though AFAIK in libcrypto so far we don't really have a concept of warnings, though would be nice to have.
What the strict cert checking (if enabled) does in case of non-compliance is to throw an error and fail.

**Comment by richsalz** [comment]
Well the SKID is not a MUST, it's a SHOULD so throwing an error when it's missing is wrong.

**Comment by DDvO** [comment]
> Well the SKID is not a MUST, it's a SHOULD so throwing an error when it's missing is wrong.

Sure, and that's why it will be hard to do it at lib level.

**Comment by richsalz** [comment]
Your plan makes sense.  But I think you should add  a config option to make it *not* happen.  A common use of openssl is to generate test certs.

**Comment by DDvO** [comment]
This brings me to a refinement of my plan:

Make the following X.509 extension configuration options the default:
```
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid
```
and provide for both of them the possibility to provide instead the value `no` for disabling (default) generation, i.e.,
```
subjectKeyIdentifier=no
authorityKeyIdentifier=no
```

**Comment by t8m** [comment]
Isn't it possible to specify empty value as in `subjectKeyIdentifier=` already? (I did not test that.)

**Comment by DDvO** [comment]
An empty value for SKID and AKID is currently not supported - for instance,
```
apps/openssl req -new -x509 -key fd.key -out fd.crt -subj "/CN=fd" -addext subjectKeyIdentifier=hash -addext authorityKeyIdentifier=
```
yields
```
Error Loading command line extensions
C0E73C60D87F0000:error:1100006C:X509 V3 routines:(unknown function):invalid empty name:crypto/x509/v3_utl.c:360:
C0E73C60D87F0000:error:11000069:X509 V3 routines:(unknown function):invalid extension string:crypto/x509/v3_conf.c:102:name=authorityKeyIdentifier,section=
C0E73C60D87F0000:error:11000080:X509 V3 routines:(unknown function):error in extension:crypto/x509/v3_conf.c:48:section=default, name=authorityKeyIdentifier, value=
```

Using an empty string rather than `no` to indicate forbidding the default value sounds attractive,
but is problematic because there are many options for which giving no value has the same effect as not giving the option at all,
which may have a different meaning. In other words, specifying `-option ""` can mean to return to the default, which is not what we want here.
So using the value `no` would be more clear than providing the empty string.

**Comment by t8m** [comment]
OK, I agree

**Comment by richsalz** [comment]
I like your refinement; thanks @ddvo.

**Comment by DDvO** [comment]
Meanwhile I've implemented this and will provide the PR when #13614 has been merged,
which cleans up the ground for it (and which I had to extend further).
After all those preparations the generation of default SKID and AKID, to be done here, becomes rather straightforward.
